source/containerimage/ocilayout.go

```diff
@@ -0,0 +1,195 @@
+func GetOCILayoutResolver(layoutPath string, sm *session.Manager, sessionID string, g session.Group) *OCILayoutResolver {
+	r := &OCILayoutResolver{
+		path:      layoutPath,
+		sm:        sm,
+		sessionID: sessionID,
+		g:         g,
+	}
+	return r
+}
+
+type OCILayoutResolver struct {
+	remotes.Resolver
+	path      string
+	sm        *session.Manager
+	sessionID string
+	g         session.Group
+	is        images.Store
+	mode      source.ResolveMode
+}
+
+// WithImageStore returns new resolver that can also resolve from local images store
+func (r *OCILayoutResolver) WithImageStore(is images.Store, mode source.ResolveMode) *OCILayoutResolver {
+	r2 := *r
+	r2.is = is
+	r2.mode = mode
+	return &r2
+}
+
+// Fetcher returns a new fetcher for the provided reference.
+func (r *OCILayoutResolver) Fetcher(ctx context.Context, ref string) (remotes.Fetcher, error) {
+	return r, nil
+}
+
+// Fetch get an io.ReadCloser for the specific content
+func (r *OCILayoutResolver) Fetch(ctx context.Context, desc ocispecs.Descriptor) (io.ReadCloser, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	sessionID := r.sessionID
+
+	caller, err := r.sm.Get(timeoutCtx, sessionID, false)
+	if err != nil {
+		return r.fetchWithAnySession(ctx, desc)
+	}
+
+	return r.fetchWithSession(ctx, desc, caller)
+}
+
+func (r *OCILayoutResolver) fetchWithAnySession(ctx context.Context, desc ocispecs.Descriptor) (io.ReadCloser, error) {
+	var rc io.ReadCloser
+	err := r.sm.Any(ctx, r.g, func(ctx context.Context, _ string, caller session.Caller) error {
+		readCloser, err := r.fetchWithSession(ctx, desc, caller)
+		if err != nil {
+			return err
+		}
+		rc = readCloser
+		return nil
+	})
+	return rc, err
+}
+
+func (r *OCILayoutResolver) fetchWithSession(ctx context.Context, desc ocispecs.Descriptor, caller session.Caller) (io.ReadCloser, error) {
+	store := sessioncontent.NewCallerStore(caller, fmt.Sprintf("oci-layout:%s", r.path))
+	readerAt, err := store.ReaderAt(ctx, desc)
+	if err != nil {
+		return nil, err
+	}
+	// just wrap the ReaderAt with a Reader
+	return io.NopCloser(&readerAtWrapper{readerAt: readerAt}), nil
+}
+
+// Resolve attempts to resolve the reference into a name and descriptor.
+// OCI Layout does not (yet) support tag name references, but does support hash references.
+func (r *OCILayoutResolver) Resolve(ctx context.Context, refString string) (string, ocispecs.Descriptor, error) {
+	ref, err := reference.Parse(refString)
+	if err != nil {
+		return "", ocispecs.Descriptor{}, fmt.Errorf("invalid reference '%s': %v", refString, err)
+	}
+	var (
+		dig  digest.Digest
+		info content.Info
+		rc   io.ReadCloser
+	)
+	if dig = ref.Digest(); dig == "" {
+		return "", ocispecs.Descriptor{}, fmt.Errorf("reference must have format @sha256:<hash>: %s", refString)
+	}
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	sessionID := r.sessionID
```

Comment:
unnecessary var

Reply:
Yeah, good point. It was an artifact of an earlier attempt that actually needed another var. Removed.